### PR TITLE
[Enhancement] Enhance mv rewrite when mv/query have multi same tables

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：
- Do enhancements for queries/mvs that have multi same tables:
  -  Because we cannot decide which tables are needed to compensate into query, so iterate the possible permutations to `tryRewrite`;
- Add `expectQueryToMVRelationIds ` for the missing tables from mv to query, so can reduce the iteration times;
- Shortcut viewDelta rewrite because view-delta only supports inner/left-outer joins; 
- Do deep copy `EquivalenceClasses ` when iterating possible `relationIdMappings `;

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 1]
```diff
@@ -163,18 +168,79 @@ public OptExpression rewrite() {
...
+            List<List<TableScanDesc>> mvExtraTableScanDescLists = Lists.newArrayList();
+            for (TableScanDesc mvTableScanDesc : mvTableScanDescs) {
+                if (!queryTableScanDescs.contains(mvTableScanDesc)) {
+                    mvExtraTableScanDescLists.add(mvTableToTableScanDecs.get(mvTableScanDesc.getTable()));
+                }
+            }
+
+            List<Set<TableScanDesc>> mvDistinctScanDescs = Lists.newArrayList();
+            PermutationGenerator generator = new PermutationGenerator(mvExtraTableScanDescLists);
+            while (generator.hasNext()) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:201]: add some comment for this core logic

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 2]
```diff
@@ -425,10 +505,21 @@ private boolean graphBasedCheck(MutableGraph<TableScanDesc> graph, List<TableSca
...
         for (Map.Entry<TableScanDesc, List<ColumnRefOperator>> entry : extraTableColumns.entrySet()) {
             int relationId = queryRefFactory.getNextRelationId();
-            for (ColumnRefOperator columnRef : entry.getValue()) {
+            List<ColumnRefOperator> columnRefOperators = entry.getValue();
+            int mvRelationId = -1;
+            for (ColumnRefOperator columnRef : columnRefOperators) {
+                if (mvRelationId == -1) {
+                    mvRelationId = mvRefFactory.getRelationId(columnRef.getId());
+                } else {
+                    Preconditions.checkState(mvRelationId == mvRefFactory.getRelationId(columnRef.getId()));
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:520]: extraTableColumns must come from the same relation. I think this check can be removed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 3]
```diff
@@ -1022,7 +1092,8 @@ protected ScalarOperator replaceExprWithTarget(ScalarOperator expr,
     private List<BiMap<Integer, Integer>> generateRelationIdMap(
             ColumnRefFactory queryRefFactory, List<Table> queryTables, OptExpression queryExpression,
             ColumnRefFactory mvRefFactory, List<Table> mvTables, OptExpression mvExpression,
-            MatchMode matchMode, Map<Table, Set<Integer>> compensationRelations) {
+            MatchMode matchMode, Map<Table, Set<Integer>> compensationRelations,
+            Map<Integer, Integer> expectQueryToMVRelationIds) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:1096]:
```suggestion
            Map<Integer, Integer> expectedExtraQueryToMVRelationIds) {
```
?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java]: fix it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:214]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/EquivalenceClasses.java [hunk 4]
```diff
@@ -33,9 +32,15 @@ public EquivalenceClasses() {
...
-        columnToEquivalenceClass = Maps.newHashMap();
-        columnToEquivalenceClass.putAll(other.columnToEquivalenceClass);
+    @Override
+    public EquivalenceClasses clone() {
+        final EquivalenceClasses ec = new EquivalenceClasses();
+        for (Map.Entry<ColumnRefOperator, Set<ColumnRefOperator>> entry :
+                this.columnToEquivalenceClass.entrySet()) {
+            ec.columnToEquivalenceClass.put(entry.getKey(), Sets.newLinkedHashSet(entry.getValue()));
+        }
+        ec.cacheColumnToEquivalenceClass = null;
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/EquivalenceClasses.java:42]: Just set it explicitly.

**Review by murphyatwork — APPROVED** [review]

**Review by ABingHuang — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**